first-party/jib-layer-filter-extension-maven/src/main/java/com/google/cloud/tools/jib/maven/extension/layerfilter/JibLayerFilterExtension.java

```diff
@@ -102,9 +122,120 @@ public ContainerBuildPlan extendContainerBuildPlan(
         .forEach(newPlanBuilder::addLayer);
+
+    ContainerBuildPlan newPlan = newPlanBuilder.build();
+
+    if (config.get().isCreateParentLayers()) {
+      newPlan = moveParentDepsToNewLayers(newPlan, mavenData, logger);
+    }
+
+    return newPlan;
+  }
+
+  private ContainerBuildPlan moveParentDepsToNewLayers(
+      ContainerBuildPlan buildPlan, MavenData mavenData, ExtensionLogger logger)
+      throws JibPluginExtensionException {
+
+    logger.log(LogLevel.LIFECYCLE, "Moving parent dependencies to new layers.");
+
+    // the key is the expected path for the parent dependency
+    Map<String, Artifact> parentDependencies =
+        getParentDependencies(mavenData)
+            .stream()
+            .map(d -> d.getArtifact())
+            // TODO: configurable?
+            .collect(
+                // TODO: getVersion or getBaseVersion? check in jib, which version it is...
+                Collectors.toMap(
+                    a -> "/app/libs/" + a.getArtifactId() + "-" + a.getVersion() + ".jar", a -> a));
+
+    // parent dependencies that have not been found in any layer (due to different version or
+    // filtering)
+    Map<String, Artifact> parentDependenciesNotFound = new HashMap<>(parentDependencies);
+
+    List<FileEntriesLayer.Builder> newLayers = new ArrayList<>();
+
+    @SuppressWarnings("unchecked")
+    List<FileEntriesLayer> originalLayers = ((List<FileEntriesLayer>) buildPlan.getLayers());
+    originalLayers.forEach(
+        l -> {
+          // for each layer, create a parent layer
+          String parentLayerName = l.getName() + "-parent";
+          FileEntriesLayer.Builder toParentLayerBuilder =
+              FileEntriesLayer.builder().setName(parentLayerName);
+          newLayers.add(toParentLayerBuilder);
+          // ... and the normal layer
+          FileEntriesLayer.Builder toLayerBuilder = FileEntriesLayer.builder().setName(l.getName());
+          newLayers.add(toLayerBuilder);
+
+          l.getEntries()
+              .forEach(
+                  fe -> {
+                    String path = fe.getExtractionPath().toString();
+                    if (parentDependencies.containsKey(path)) {
+                      // move to parent layer
+                      logger.log(LogLevel.INFO, "Moving " + path + " to " + parentLayerName + ".");
+                      toParentLayerBuilder.addEntry(fe);
+                      // mark parent dep as found
+                      parentDependenciesNotFound.remove(path);
+                    } else {
+                      // keep in original layer
+                      logger.log(LogLevel.INFO, "Keep " + path + " in " + l.getName() + ".");
+                      toLayerBuilder.addEntry(fe);
+                    }
+                  });
+        });
+
+    parentDependenciesNotFound.forEach(
+        (path, artifact) -> {
+          logger.log(LogLevel.ERROR, "Dependency from parent not found: " + path);
+        });
```

Comment:
```suggestion
        (path, artifact) -> logger.log(LogLevel.ERROR, "Dependency from parent not found: " + path));
```
When this can happen? And if it's an ERROR, should we stop and throw an exception?

Reply:
Oh, maybe this is when the files have been removed by the filters (by whatever other extensions)? I think `LogLevel.INFO` sounds reasonable. (I thought about `WARN`, but usually that's for something actionable to fix the situation.)

Reply:
@chanseokoh , I see the following possible causes here:

- As you said, it could have been removed by a filter before. INFO is fine with me.
- I could also have been removed by some sort of Maven dependeny exclusion, altough I am not sure right know, if exclusion is possible for dependencies from parent. INFO ist okay, nevertheless.
- The version of the parent dependency could have been overriden with a new version. I would like to give a WARN in this case, since this is bad practice anyway and would also negate the intended effect of a common parent layer. As sugar I would like to find an log the version that has been found for this artifact id (which was the reason to use a Map in the first place)

Reply:
> The version of the parent dependency could have been overridden with a new version. I would like to give a WARN in this case, since this is bad practice anyway and would also negate the intended effect of a common parent layer.

I think either INFO and WARN makes sense depending on how we look at it and how the extension eventually works. It's a bad practice that this extension could encourage to fix, but the extension should be able to correctly isolate this case without false positives. (I know some users are forced to remove all warnings if there are any, and false positives are really bad.) And in some sense, it may not really be the business of the extension to worry about the bad practice; the job of the extension is to just move or not move the dependency. The worst case is just that the dependency will be either in the original layer or the parent layer, which may not match what the user expected; nothing actually breaks at runtime. Even if the dependency is not moved, it will only be a minor inefficiency to the user if that's what the user didn't want. But yeah, in some other sense, because this extension is in the business of moving parent dependencies, it may make sense to warn this case (if it can be correctly identified).

But most importantly, I worry that it may be possible that this warning case is not something the user can take an action on. For example, I believe Maven resolves version conflicts by choosing the one with the least depth in the dependency graph. Then, for example, a parent dependency transitively brings in foo-1.0.0.jar, the user doesn't have control over the parent POM, another dependency in a sub-module brings in foo-2.0.0.jar transitively (or maybe it's just a direct dependency), and Maven chooses foo-2.0.0.jar because it's closer than 1.0.0 from the root of the dep graph.

Reply:
Ok, I change to INFO. I think you are right that this extension is not the right place to care or warn about version conflicts or similar...

Reply:
For the moment, I kept the Map and the additional braces here, since I still intend to enhance the logging if I can find a different version of the artifact.